### PR TITLE
feat: Day.render / CalendarDays.render

### PR DESCRIPTION
**Pull request by astahmer**
Hey, I needed to customize the content of each day's cell, this should cover it

example story:
Screenshot 2022-06-01 at 01 00 39

with: 
```ts
<CalendarDays
            render={args =>
              new Date(args.day).getDate() < 8 ? (
                <Box d="flex" flexDirection="column" alignItems="center">
                  <div>{format(args.day, 'd')}</div>
                  <Circle size="4px" bgColor="pink.300" />
                </Box>
              ) : (
                format(args.day, 'd')
              )
            }
          />
```

### CODE REVIEW

**Comment by uselessdev** [comment]
Hi, @astahmer first of all thank you for opening this pull request.

So, about the implementation, I really want to avoid the use of render props _(yes I know I'm using render props in controls but I'm planning to change this)_ and use `children` instead. I suggest something like this:

```tsx
<CalendarDays>
  {({ day }) => {
    if (new Date(day).getDate() < 8) {
      return (
        <Box d="flex" flexDirection="column" alignItems="center">
          <div>{day /* should use date-fns format as well */}</div>
          <Circle size="4px" bgColor="pink.300" />
        </Box>
      )
    }

    return day
  }}
</CalendarDays>
```

But, I still don't know if this is 'ok' maybe we can find another solution something like exposing a hook such as `useCalendarDays` and:

```tsx
const { days } = useCalendarDays()

<CalendarDays>
  {days.map(day => <Box>{day}</Box>)}
</CalendarDays>
```

**Comment by astahmer** [comment]
Ok I see, no problem i'll change it ! 👍  Out of curiosity, is there a problem with render props ?

I think that with a hook such as `const { days } = useCalendarDays()`, you'd be missing the built-in logic (`isSelected`/`isInRange`/`isDisabled`), so I would not really consider that option.

IMO you could make it a hook at the Day level:
```ts
`const { day, isSelected, isInRange, isDisabled, variant } = useCalendarDay()`
```

or make it another compound component at the Day level:
```ts
<CalendarDays>
    <CalendarDay>
        {({ day }) => (day ? `(${format(day, 'd')})` : <span />)}
    </CalendarDay>
<CalendarDays>
```

or even better we can just make both and `useCalendarDay` inside `CalendarDay` so that the user has the same level of control as the built-in components

**Comment by uselessdev** [comment]
> Ok I see, no problem i'll change it ! 👍 Out of curiosity, is there a problem with render props ?

absolutely not

> or make it another compound component at the Day level:

I was thinking of something like this

File: src/calendar.stories.tsx [hunk 1]
```diff
@@ -773,18 +774,128 @@ export const WithCustomDayRender: ComponentStory<typeof Calendar> = () => {
...
+          <div>{format(day, 'd')}</div>
+          <Circle size="4px" bgColor="pink.300" />
+        </Box>
+      ) : (
+        format(day, 'd')
+      )}
+    </Day>
+  )
+}
+export const CustomDayComponent: ComponentStory<typeof Calendar> = () => {
```
[the remark this answers is not in the document]
Comment (uselessdev) [reply src/calendar.stories.tsx:881]: For the story I was thinking in something like this:
```tsx
function CustomDay({ day, ...props }: Day) {
  if (new Date(day).getDate() > 8) {
    return <CalendarDay day={day} {...props} />
  }

  return (
    <CalendarDay
      day={day}
      {...props}
      flexDir="column"
      rounded="md"
      _hover={{ bgColor: 'pink.100' }}
    >
      {format(day, 'd')}
      <Circle size="4px" bgColor="pink.300" />
    </CalendarDay>
  )
}

export const WithCustomDay: ComponentStory<typeof Calendar> = () => {
  const [dates, setDates] = React.useState<CalendarValues>({})

  const handleSelectDate = (dates: CalendarValues) => setDates(dates)

  return (
    <Calendar value={dates} onSelectDate={handleSelectDate}>
      <CalendarControls>
        <CalendarPrevButton />
        <CalendarNextButton />
      </CalendarControls>

      <CalendarMonths>
        <CalendarMonth>
          <CalendarMonthName />
          <CalendarWeek />
          <CalendarDays>{props => <CustomDay {...props} />}</CalendarDays>
        </CalendarMonth>
      </CalendarMonths>
    </Calendar>
  )
}
```

this is similar to what you already did

**Review by uselessdev — CHANGES_REQUESTED** [review]

**Comment by astahmer** [comment]
> we can keep this as a `type` and we can use `ButtonProps` to have style props in the `CalendarDay`
> if we add the ButtonProps in Day type we will need to spread the props after the children something like this:
> and we can now merge with styles in the sx props on Button component

done in the last commit 👍 

> Now this component will be 'public' we should rename it to CalendarDay

that makes sense but at the same time, then how would you name the built-in component containing the `isSelected` / `isDisabled` / etc logic (either directly in the render fn or in the future using the `useCalendarDay` hook) ? 
or does that mean there would not be one provided ?
like [this one](https://github.com/uselessdev/datepicker/blob/556da33bd35bff7b675dd35763960c8eaffcf5ce/src/month-days.tsx#L46) 

> this file has too much change, for this PR I want to suggest only add props for the children:
> let's keep this simple by only adding the WithCustomDay story.
> I really like the hook, but I think we should add this in another pull request.

if you don't mind, could you do it ? this might be more straightforward than me doing it since i'm not really sure how you want to split the PR

**Review by uselessdev — APPROVED** [review]

**Comment by uselessdev** [comment]
@all-contributors please add @astahmer for code

**Pull request merged**